uber-go/cadence-client
# Rename .gen -> gen for go 1.16 compatibility

Client-side counterpart to uber/cadence/issues/3987 .  This essentially has to be merged and released before the server can be upgraded.

In a nutshell: `go` hasn't liked `.gen` since some time in 2018, but 1.16 was just released and now it's enforced.
Doing anything with 1.16 with modules enabled produces errors like:
```
go.uber.org/cadence/workflow imports
go.uber.org/cadence/.gen/go/shared: malformed import path "go.uber.org/cadence/.gen/go/shared": leading dot in path element
```

So this commit:
- Pins tool versions via tools.go + go.mod, as builds did not appear stable before (likely transitive thrift dependencies updating, breaking generated code)
- Changes the generated source dirs from ".gen" to "gen"
- Changes licensegen to look at the new path
- Find-and-replaces all `go.uber.org/cadence/{.gen,gen}`
- To test:
    - In this repo: clean and rebuild / thriftc / `go test` with both go 1.15 and go 1.16, and make sure it's stable (it is, minus a very weird shifting import in generated code.  goimports would fix that.)
    - In a test repo that imports this repo: do the same.  Build, test, `go mod` stuff.  This failed earlier, works now.

Implications of these changes include:
- This _is_ a breaking change.  Unfortunately the `.gen` imports are exposed directly (via e.g. `workflow/error.go` -> `.gen/go/shared.TimeoutType` argument) and transitively (e.g. `internal/workflow.go` -> `WorkflowInfo` contains `.gen/go/shared.Memo`), so making this non-breaking may be quite a lot of work.
    - It is however _possible_: duplicate every .gen-exposing file and rewrite the imports in one copy, and use `// +build go1.16` tags to control compilation.  I'm not sure how far this goose would have to be chased to get it all done, but I can look closer if desired.
- Some build and runtime dependencies have changed versions in relatively minor ways.  This mostly brings them more in line with the versions used to generate code.
    - I have not tried to upgrade any back to the previous versions, but it would be easy to try.
    - Generated code appears identical.
- Users may end up downloading more dependencies (e.g. staticcheck), maybe not, depending on how Go decides to handle non-buildable transitive dependencies.  They've always _intended_ to prune them implicitly, but I do not know if that is happening yet.
    - They do not appear to be _constrained_ by these dependencies, however, as they have no actual import, and do not appear in their go.mod files.  Upgrading or downgrading in a test project worked and did not require `replace` directives.
- For good or bad, transitive dependencies of tools and this library are now bound together.
    - This is pretty common and not usually an issue, but if it becomes one: we can switch conflicting tools to a separate tools.mod and deal with the internal-only quirks that causes.
    - I can do that ^ right now if we want to be paranoid, and it'd eliminate the chance of tool versions impacting users, but it opens the door to version skew.  Neither option is ideal ¯\_(ツ)_/¯
